hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerReportForDuty.java

```diff
@@ -241,6 +243,125 @@ public void run() {
+    assertEquals(1, master.getMaster().getServerManager().getOnlineServers().size());
+
+    // Decommission the region server and tries to re-add it
+    List<ServerName> serversToDecommission = new ArrayList<ServerName>();
+    serversToDecommission.add(rs.getRegionServer().getServerName());
+    master.getMaster().decommissionRegionServers(serversToDecommission, true);
+
+    // Assert that the server is now decommissioned
+    ServerName decommissionedServer = master.getMaster().listDecommissionedRegionServers().get(0);
+    assertEquals(1, master.getMaster().listDecommissionedRegionServers().size());
+    assertEquals(1, master.getMaster().getServerManager().getDrainingServersList().size());
+    assertEquals(1, master.getMaster().getServerManager().getOnlineServers().size());
+    assertEquals(rs.getRegionServer().getServerName().toString(),
+      decommissionedServer.getServerName());
+
+    // Create a second region server
+    cluster.getConfiguration().set(HConstants.REGION_SERVER_IMPL, MyRegionServer.class.getName());
+    rs2 = cluster.addRegionServer();
+    rs2.start();
+    waitForSecondRsStarted();
+
+    // Assert that the number of decommissioned and live hosts didn't change and that the hostname
+    // of rs2 matches that of the decommissioned server
+    String rs2HostName = rs2.getRegionServer().getServerName().getHostname();
+    assertEquals(1, master.getMaster().listDecommissionedRegionServers().size());
+    assertEquals(1, master.getMaster().getServerManager().getDrainingServersList().size());
+    assertEquals(1, master.getMaster().getLiveRegionServers().size());
+    assertEquals(1, master.getMaster().getServerManager().getOnlineServers().size());
+    assertEquals(rs2HostName, decommissionedServer.getHostname());
+  }
+
+  /**
+   * Tests that the RegionServer's reportForDuty gets accepted by the master when the master is not
+   * configured to reject decommissioned hosts, even when there is a match for the joining
+   * RegionServer in the list of decommissioned servers. Test case for HBASE-28342.
+   */
+  @Test
+  public void testReportForDutyGetsAcceptedByMasterWhenNotConfiguredToRejectDecommissionedHosts()
+    throws Exception {
+    // Start a master and wait for it to become the active/primary master.
+    // Use a random unique port
+    cluster.getConfiguration().setInt(HConstants.MASTER_PORT, HBaseTestingUtil.randomFreePort());
+    cluster.getConfiguration().setInt(ServerManager.WAIT_ON_REGIONSERVERS_MINTOSTART, 1);
+    cluster.getConfiguration().setInt(ServerManager.WAIT_ON_REGIONSERVERS_MAXTOSTART, 1);
+
+    // Set the cluster to not reject decommissioned hosts (default behavior)
+    cluster.getConfiguration().setBoolean(HConstants.REJECT_DECOMMISSIONED_HOSTS_KEY, false);
+
+    master = cluster.addMaster();
+    rs = cluster.addRegionServer();
+    LOG.debug("Starting master: " + master.getMaster().getServerName());
+    master.start();
+    rs.start();
+
+    waitForClusterOnline(master);
```

Comment:
Does this wait for the rs to join the cluster, or only wait for the master to come online? Or maybe a single rs being present a requirement for the master to be online?

Reply:
This method call waits for the master to initialize and for the `rs` instance variable thread to start.